selection/utils.py

```diff
@@ -55,3 +59,120 @@ def get_utilized_indexes(
             }
 
     return utilized_indexes_workload, query_details
+
+
+# --- Join Order Benchmark utilities ---
+
+IMDB_LOCATION = "https://archive.org/download/imdb_20200624/imdb.zip"
+IMDB_FILE_NAME = "imdb.zip"
+IMDB_TABLE_DIR = "imdb_data"
+IMDB_TABLE_NAMES = [
+    "aka_name",
+    "aka_title",
+    "cast_info",
+    "char_name",
+    "company_name",
+    "company_type",
+    "comp_cast_type",
+    "complete_cast",
+    "info_type",
+    "keyword",
+    "kind_type",
+    "link_type",
+    "movie_companies",
+    "movie_info",
+    "movie_info_idx",
+    "movie_keyword",
+    "movie_link",
+    "name",
+    "person_info",
+    "role_type",
+    "title",
+]
+
+
+def _clean_up(including_table_dir=False):
+    if os.path.exists(IMDB_FILE_NAME):
+        os.remove(IMDB_FILE_NAME)
+
+    if including_table_dir and os.path.exists(IMDB_TABLE_DIR):
+        for file in os.listdir(IMDB_TABLE_DIR):
+            os.remove("./%s/%s" % (IMDB_TABLE_DIR, file))
+        os.rmdir(IMDB_TABLE_DIR)
+
+
+def _files_exist():
+    for table_name in IMDB_TABLE_NAMES:
+        if not os.path.exists(os.path.join(IMDB_TABLE_DIR, table_name + ".csv")):
+            return False
+
+    return True
+
+
+def download_and_uncompress_imdb_data():
+    if _files_exist():
+        logging.info("IMDB already present.")
+        return True
+
+    logging.critical("Retrieving the IMDB dataset - this may take a while.")
+
+    # We are going to calculate the md5 hash later, on-the-fly while downloading
+    hash_md5 = hashlib.md5()
+
+    url = urllib.request.urlopen(IMDB_LOCATION)
+    meta = url.info()
+    file_size = int(meta["Content-Length"])
+
+    file = open(IMDB_FILE_NAME, "wb")
+
+    logging.info(f"Downloading: {IMDB_FILE_NAME} ({b_to_mb(file_size):.3f} MB)")
+
+    already_retrieved = 0
+    block_size = 8192
+    try:
+        while True:
+            buffer = url.read(block_size)
+            if not buffer:
+                break
+
+            hash_md5.update(buffer)
+
+            already_retrieved += len(buffer)
+            file.write(buffer)
+            status = (
+                f"Retrieved {already_retrieved * 100.0 / file_size:3.2f}% of the data"
+            )
+            status = f"{status}{chr(8) * (len(status) + 1)}"
```

Comment:
What is the `chr(8) * (len(status) + 1)` doing?

Reply:
I added a comment: "chr(8) refers to a backspace. In conjunction with end="\r", this overwrites the previous status value and achieves the right padding."